lending: Use fair obligation health factor calculation

jstarry opened the pull request:
#### Problem
Loans that are large in value relative to the depth of the dex order book are prone to having their collateral devalued which leads to a higher LTV ratio and potential opportunity for liquidations.

#### Changes
- Use the best price in the order book to determine the health of an obligation but continue using order book depth to calculate a fair market value of of collateral to send to liquidators. This leaves large obligations much less vulnerable to order book manipulation and provides an efficient way to determine the health of a loan. An additional win is that the amount of liquidity a liquidator wants to pay off has no impact on the determination about an obligation's health.
- Add special handling for liquidating dust obligations. When an obligation has less than a single token unit of borrowed value remaining, allow a liquidator to pay it off in return for the remaining collateral.

joncinque commented on token-lending/program/src/state/reserve.rs:
```diff
@@ -95,41 +95,46 @@ impl Reserve {
-            return Err(LendingError::ObligationTooSmall.into());
+
+        // Special handling for dust obligations
+        if max_repayable == 0 {
```
Were you able to check that the proptests hit this case?  If not, could you add a special test case for this?

joncinque replied:
Also, can this happen in normal repays too?

jstarry replied:
Yes, there's already handling for normal repays for this case. I'll add some explicit test cases

joncinque reviewed (approved):
Looks good!  Just two small questions really, which don't need to be done in this PR.

jstarry commented:
@joncinque do you mind taking a look at the latest changes? In writing your requested tests I found some holes.
- Repay amount for dust accounts was hard coded to `1` but should be the ceiling of the decimal borrow amount
- If decimal borrow amount was between 1 and 2, we weren't enforcing that the input liquidity amount was greater than 1.
- There were some edge cases where unhealthy loans couldn't be liquidated, I fixed those and refactored the test to ensure that all unhealthy loans can be liquidated.
- Decided that splitting off the bonus amount wasn't worth the compute cost and by combining the bonus and withdraw amounts, we're now giving liquidators a more fair amount of collateral (rather than rounding down separately)

jstarry commented on token-lending/program/src/dex_market.rs:
```diff
@@ -59,6 +59,43 @@ pub struct TradeSimulator<'a> {
+            TradeAction::Sell
+        };
+
+        let currency = if token_mint == self.quote_token_mint {
+            Currency::Quote
+        } else {
+            Currency::Base
+        };
+
+        let order_book_side = match (action, currency) {
+            (TradeAction::Buy, Currency::Base) => Side::Ask,
+            (TradeAction::Sell, Currency::Quote) => Side::Ask,
+            (TradeAction::Buy, Currency::Quote) => Side::Bid,
+            (TradeAction::Sell, Currency::Base) => Side::Bid,
+        };
+        if order_book_side != self.orders_side {
+            return Err(LendingError::DexInvalidOrderBookSide.into());
+        }
```
will do later

joncinque commented:
Looks really good, just some questions and nits

jstarry commented on token-lending/program/src/state/reserve.rs:
```diff
@@ -795,84 +788,102 @@ mod test {
-            liquidation_bonus in 0..=100u8,
-            liquidation_threshold in 2..=100u8,
+        fn unhealthy_obligations_can_be_liquidated(
+            obligation_collateral in 1..=u32::MAX,
```
Yes, fixed

jstarry commented on token-lending/program/src/state/reserve.rs:
```diff
@@ -92,85 +95,90 @@ impl Reserve {
+        let max_repay_amount = obligation
+            .borrowed_liquidity_wads
+            .try_mul(Rate::from_percent(LIQUIDATION_CLOSE_FACTOR))?
+            .try_floor_u64()?;
```
Yes, not sure how that happened.

jstarry reviewed (commented):
Thanks for the detailed feedback @joncinque!

jstarry commented:
feedback changes can be found here: https://github.com/solana-labs/solana-program-library/commit/ca5e43f3c28b205a8c10342d47ce5a33a75de925

joncinque commented on token-lending/program/src/state/reserve.rs:
```diff
@@ -92,85 +95,90 @@ impl Reserve {
-        // Special handling for dust obligations
-        if max_repayable == 0 {
+        // Special handling for small, closeable obligations
+        let close_amount = liquidity_amount.min(CLOSEABLE_AMOUNT);
```
Right, of course, I forgot about the 50% max liquidation!  Ok great, thanks for the explanation

joncinque commented:
Looks good
